# Satisfaction, Loyalty and its drivers by client segment

## Background

Customer satisfaction plays an important role in building strong client relationships, encouraging loyalty and supporting long-term business growth. Understanding how clients view the company’s services can help identify areas of strength and opportunities for improvement.

This analysis examines customer satisfaction across different client segments and years. It considers the quality of the company’s services as well as clients’ willingness to recommend the company.

## Analysis Overview

The analysis focuses on three client segments:

* **Government organisations (Govt)**
* **Non-Profit Organisations (NPO)**
* **Private organisations**

Customer satisfaction is evaluated using four service dimensions:

* Presales & Partnership
* Technical Expertise
* Project Delivery
* Post-Sales Support

Each service dimension is rated from **1 to 5**, where a higher value represents greater customer satisfaction.

The analysis also uses **NPS Rating**, which is recorded from **0 to 10**. NPS represents a client’s willingness to recommend the company and is used as an indication of customer advocacy, loyalty and overall experience. A higher NPS Rating generally suggests a more positive client relationship.

As the dataset contains individual NPS ratings, this analysis uses the **average NPS Rating** instead of calculating the formal Net Promoter Score.

## Purpose of the Analysis

The purpose of this analysis is to:

1. Compare customer satisfaction across different client segments.
2. Examine how satisfaction and NPS ratings change over time.
3. Identify client segments with relatively high or low satisfaction.
4. Determine which service dimensions are most strongly associated with NPS Rating.
5. Highlight areas where service improvements may strengthen customer satisfaction and loyalty.

## Connection to the Main Problem

The team’s main objective is to determine which client segments generate the highest gross profit while maintaining strong customer satisfaction.

This analysis provides the **customer satisfaction component** of the main problem. Its findings can be combined with the profitability analysis to identify client segments that provide both strong financial value and positive customer outcomes.



## Data preparation and validation

The analysis uses the cleaned, merged client dataset. Each row represents a **client-year observation**, so the same client may legitimately appear in different years.

- Service dimensions are rated on a **1–5 scale**.
- NPS Rating is recorded on a **0–10 scale**.
- Client Type is the main segmentation variable used in these visualisations.

The checks below confirm the dataset shape, inspect the available fields and verify that there is no more than one record for the same Client ID and Year combination.

In [10]:
import pandas as pd
import plotly.express as px

df = pd.read_excel("merged_cleaned.xlsx")

print("Shape:", df.shape)
print(
    "Duplicate client-year records:",
    df.duplicated(subset=["CLIENT ID", "YEAR"]).sum()
)

display(df.head())

Shape: (392, 16)
Duplicate client-year records: 0


,CLIENT ID,TYPE,COMMENCEMENT DATE,STAFF STRENGTH,SECTOR,COUNTRY,YEAR,PRESALES AND PARTNERSHIP,TECHNICAL EXPERTISE,PROJECT DELIVERY,POST-SALES SUPPORT,NPS RATING,NPS RATING.1,NPS RATING.2,SOFTWARE,MANPOWER
0,G1-001,Govt,2020-01-19 00:00:00,> 200,Finance,Singapore,2021,3,4,4,5,8,148180,17321,25562,41201
1,G1-001,Govt,2020-01-19 00:00:00,> 200,Finance,Singapore,2022,5,5,3,4,9,197313,23604,32875,50085
2,G1-001,Govt,2020-01-19 00:00:00,> 200,Finance,Singapore,2023,5,4,3,4,7,349899,51332,71695,89089
3,G1-001,Govt,2020-01-19 00:00:00,> 200,Finance,Singapore,2024,5,5,3,4,8,268821,40969,56452,76545
4,G1-001,Govt,2020-01-19 00:00:00,> 200,Finance,Singapore,2025,5,4,5,5,10,138115,14516,20889,28968


In [11]:
print(
    "Duplicate client-year records:",
    df.duplicated(subset=["CLIENT ID", "YEAR"]).sum()
)

Duplicate client-year records: 0


### Validation result

The cleaned dataset contains **392 client-year observations and 16 variables**, with **0 duplicate Client ID–Year records**. This supports fair yearly comparisons without double-counting the same client within a year.

The next two cells define the four service dimensions and calculate an initial client-type summary before the interactive visualisations are created.


In [12]:
service_columns = [
    "PRESALES AND PARTNERSHIP",
    "TECHNICAL EXPERTISE",
    "PROJECT DELIVERY",
    "POST-SALES SUPPORT"
]

In [13]:
type_summary = (
    df.groupby("TYPE")[service_columns]
      .mean()
      .round(2)
      .reset_index()
)

display(type_summary)

,TYPE,PRESALES AND PARTNERSHIP,TECHNICAL EXPERTISE,PROJECT DELIVERY,POST-SALES SUPPORT
0,Govt,4.09,3.94,4.35,3.91
1,NPO,3.76,3.88,4.26,4.12
2,Private,3.96,3.92,4.01,3.80


## Visualisation 1 — Service Ratings by Client Type and Period

### Purpose

This visualisation compares the average service ratings of Government, NPO and Private clients from 2021 to 2025.

The service ratings represent clients’ satisfaction with four areas of the company’s service:

* **Presales & Partnership:** The quality of engagement before a project begins and the company’s ability to develop a strong working relationship with the client.
* **Technical Expertise:** The client’s perception of the company’s technical knowledge, professional skills and ability to provide suitable solutions.
* **Project Delivery:** How effectively the company plans, manages and completes its projects.
* **Post-Sales Support:** The quality of assistance provided after project completion, including ongoing support and issue resolution.

Each dimension is rated from **1 to 5**, where a higher rating represents greater satisfaction. The bars show the **average rating** given by clients within each client type and period.

### What the Visualisation Shows

The grouped bars allow the four service dimensions to be compared within each client type. This helps identify:

* the strongest and weakest service areas for each client segment;
* whether different client types have different service experiences;
* whether satisfaction improves or declines over time; and
* which service areas may require greater attention.

The **Overall** view summarises all years, while the yearly views show how the ratings change between 2021 and 2025.

### How to Interact

* Press **Play** to animate the changes from 2021 to 2025 and the Overall view.
* Drag the **period slider** to select a specific year.
* Hover over a bar to view the client type, service dimension, average rating and number of records.
* Click a service dimension in the legend to show or hide it.

The y-axis remains fixed from **0 to 5** so that ratings can be compared fairly across different years and client types.

### Interpretation

A taller bar represents a higher average level of satisfaction with that service dimension. Large differences between the bars may indicate that clients are more satisfied with some areas of service than others.

These comparisons help the company recognise service strengths that should be maintained and weaknesses that could be improved to support stronger customer satisfaction and loyalty.



In [14]:
import pandas as pd
import plotly.express as px


# -------------------------------------------------
# 1. Settings
# -------------------------------------------------

service_columns = [
    "PRESALES AND PARTNERSHIP",
    "TECHNICAL EXPERTISE",
    "PROJECT DELIVERY",
    "POST-SALES SUPPORT"
]

service_label_map = {
    "PRESALES AND PARTNERSHIP": "Presales & Partnership",
    "TECHNICAL EXPERTISE": "Technical Expertise",
    "PROJECT DELIVERY": "Project Delivery",
    "POST-SALES SUPPORT": "Post-Sales Support"
}

service_order = list(service_label_map.values())
type_order = ["Govt", "NPO", "Private"]

service_colors = {
    "Presales & Partnership": "#636EFA",
    "Technical Expertise": "#EF553B",
    "Project Delivery": "#00CC96",
    "Post-Sales Support": "#AB63FA"
}


# -------------------------------------------------
# 2. Calculate overall averages
# -------------------------------------------------

overall_summary = (
    df.groupby("TYPE", as_index=False)[service_columns]
      .mean()
      .round(2)
)

overall_counts = (
    df.groupby("TYPE")
      .size()
      .rename("Records")
      .reset_index()
)

overall_summary = overall_summary.merge(
    overall_counts,
    on="TYPE"
)

overall_long = overall_summary.melt(
    id_vars=["TYPE", "Records"],
    value_vars=service_columns,
    var_name="Service Dimension",
    value_name="Average Rating"
)

overall_long["Service Dimension"] = (
    overall_long["Service Dimension"]
    .replace(service_label_map)
)

overall_long["Period"] = "Overall"


# -------------------------------------------------
# 3. Calculate yearly averages
# -------------------------------------------------

yearly_summary = (
    df.groupby(["YEAR", "TYPE"], as_index=False)[service_columns]
      .mean()
      .round(2)
)

yearly_counts = (
    df.groupby(["YEAR", "TYPE"])
      .size()
      .rename("Records")
      .reset_index()
)

yearly_summary = yearly_summary.merge(
    yearly_counts,
    on=["YEAR", "TYPE"]
)

yearly_long = yearly_summary.melt(
    id_vars=["YEAR", "TYPE", "Records"],
    value_vars=service_columns,
    var_name="Service Dimension",
    value_name="Average Rating"
)

yearly_long["Service Dimension"] = (
    yearly_long["Service Dimension"]
    .replace(service_label_map)
)

yearly_long["Period"] = (
    yearly_long["YEAR"]
    .astype(int)
    .astype(str)
)


# -------------------------------------------------
# 4. Combine yearly and overall results
# -------------------------------------------------

animation_data = pd.concat(
    [
        yearly_long.drop(columns="YEAR"),
        overall_long
    ],
    ignore_index=True
)

animation_data["Animation Group"] = (
    animation_data["TYPE"]
    + " — "
    + animation_data["Service Dimension"]
)

period_order = [
    str(int(year))
    for year in sorted(df["YEAR"].unique())
] + ["Overall"]


# -------------------------------------------------
# 5. Create animated chart
# -------------------------------------------------

fig = px.bar(
    animation_data,
    x="TYPE",
    y="Average Rating",
    color="Service Dimension",
    barmode="group",

    animation_frame="Period",
    animation_group="Animation Group",

    text="Average Rating",
    custom_data=["Records"],

    category_orders={
        "Period": period_order,
        "TYPE": type_order,
        "Service Dimension": service_order
    },

    color_discrete_map=service_colors,

    labels={
        "TYPE": "Client Type",
        "Average Rating": "Average Rating"
    },

    range_y=[0, 5.4],

    title="Service Ratings by Client Type: 2021–2025 and Overall"
)


# -------------------------------------------------
# 6. Hover and data-label formatting
# -------------------------------------------------

trace_format = dict(
    texttemplate="%{y:.2f}",
    textposition="outside",
    cliponaxis=False,

    hovertemplate=(
        "<b>Client Type: %{x}</b><br>"
        "Service: %{fullData.name}<br>"
        "Average Rating: %{y:.2f} / 5<br>"
        "Records Used: %{customdata[0]}"
        "<extra></extra>"
    )
)

fig.update_traces(**trace_format)

# Apply the same formatting to every animation frame
for frame in fig.frames:
    for trace in frame.data:
        trace.update(**trace_format)


# -------------------------------------------------
# 7. Adjust animation speed
# -------------------------------------------------

play_settings = (
    fig.layout.updatemenus[0]
       .buttons[0]
       .args[1]
)

play_settings["frame"] = {
    "duration": 1400,
    "redraw": True
}

play_settings["transition"] = {
    "duration": 700,
    "easing": "cubic-in-out"
}

play_settings["fromcurrent"] = True

fig.layout.sliders[0].currentvalue.prefix = "Period: "
fig.layout.sliders[0].pad.t = 50


# -------------------------------------------------
# 8. Final formatting
# -------------------------------------------------

fig.update_layout(
    template="plotly_white",

    title=dict(
        x=0.5,
        xanchor="center"
    ),

    xaxis_title="Client Type",
    yaxis_title="Average Rating",

    legend_title="Service Dimension",

    transition=dict(
        duration=700,
        easing="cubic-in-out"
    ),

    margin=dict(
        t=100,
        b=130
    )
)

fig.update_yaxes(
    range=[0, 5.4],
    dtick=1
)


# -------------------------------------------------
# 9. Display graph
# -------------------------------------------------

fig.show(
    config={
        "displayModeBar": True,
        "displaylogo": False,
        "scrollZoom": True,
        "responsive": True,

        "toImageButtonOptions": {
            "format": "png",
            "filename": "service_ratings_story",
            "scale": 2
        }
    }
)

### Key Findings from Visualisation 1

#### Overall Performance

Average service ratings range from **3.76 to 4.35 out of 5**, indicating generally positive satisfaction across all client types.

**Project Delivery is the strongest overall dimension** for every segment:

* Government: **4.35**
* NPO: **4.26**
* Private: **4.01**

It also reaches its highest rating for all segments in **2025**: Government 4.75, NPO 4.54 and Private 4.18. This suggests that Project Delivery has strengthened over time and should remain a key service priority.

#### Trends by Client Type (2021 to 2025)

| Client Type    | Main Trend                                                                                                                                                   |
| -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| **Government** | Project Delivery improves strongly from 3.67 in 2021 to 4.75 in 2025. Presales also improves overall, while Post-Sales Support remains relatively unchanged. |
| **NPO**        | Project Delivery increases from 4.00 to 4.54. Post-Sales Support improves after 2021, but Presales & Partnership falls from 4.50 to 3.85.                    |
| **Private**    | Ratings are more stable. Technical Expertise and Project Delivery improve, while Post-Sales Support remains the weakest dimension at 3.83 in 2025.           |

Overall, **Project Delivery shows the clearest positive trend** across all three client types. Technical Expertise also improves among Government and Private clients, while Presales and Post-Sales Support show less consistent changes.

#### Business Implications

The company should:

1. Maintain its strong Project Delivery performance.
2. Review the practices used in 2025, when Project Delivery reached its highest ratings.
3. Strengthen Presales & Partnership for NPO clients.
4. Improve Post-Sales Support for Private clients.
5. Apply different improvement strategies to each client type instead of using one general approach.

#### Limitations

The chart displays average ratings, which may hide differences between individual clients. The client segments also contain different numbers of records, meaning that changes in smaller groups may be influenced by only a few responses.

The trends show how ratings changed over time but do not explain what caused those changes.



## Visualisation 2 — Average NPS Rating by Client Type and Year

### Purpose

This heatmap compares the average NPS Rating of Government, NPO and Private clients from 2021 to 2025, together with an Overall average.

NPS Rating represents a client’s willingness to recommend the company. It is used as an indicator of customer advocacy, loyalty and overall experience. A higher average rating suggests that clients are more likely to view and recommend the company positively.

### What the Visualisation Shows

Each heatmap cell represents the average NPS Rating for one client type and period. The visualisation helps identify:

* which client type has the highest or lowest average NPS Rating;
* whether customer advocacy improves or declines over time;
* years with unusually high or low ratings; and
* client segments that may require additional attention.

The **Overall** column summarises the average across all years, while the yearly columns show how NPS Rating changes from 2021 to 2025.

### How to Interpret the Heatmap

The colour of each cell represents the average NPS Rating:

* Higher values indicate stronger customer advocacy and loyalty.
* Lower values may indicate weaker advocacy or areas requiring improvement.
* The printed value allows the exact average to be compared across cells.

The colour scale focuses on the range of ratings found in the dataset. Therefore, the colour differences may appear larger than the actual numerical differences, and the printed values should always be considered.

### How to Interact

Hover over a cell to view:

* the client type;
* the selected year or Overall period;
* the average NPS Rating; and
* the number of client-year records used to calculate the average.

The record count is important because averages based on fewer responses may be more affected by individual ratings.

> **Measurement note:** This chart uses the average individual **NPS Rating** on a 0–10 scale. It does not show the formal Net Promoter Score, which is calculated as the percentage of Promoters minus the percentage of Detractors.


In [15]:
import pandas as pd
import plotly.graph_objects as go


# -------------------------------------------------
# 1. Set display order
# -------------------------------------------------

type_order = [
    client_type
    for client_type in ["Govt", "NPO", "Private"]
    if client_type in df["TYPE"].unique()
]

years = sorted(
    df["YEAR"].dropna().astype(int).unique()
)


# -------------------------------------------------
# 2. Calculate average NPS rating
# -------------------------------------------------

nps_heatmap = (
    df.groupby(["TYPE", "YEAR"])["NPS RATING"]
      .mean()
      .round(2)
      .unstack("YEAR")
      .reindex(type_order)
)

nps_heatmap.columns = [
    str(int(year))
    for year in nps_heatmap.columns
]

nps_heatmap["Overall"] = (
    df.groupby("TYPE")["NPS RATING"]
      .mean()
      .round(2)
      .reindex(type_order)
)


# -------------------------------------------------
# 3. Calculate number of records for hover
# -------------------------------------------------

record_heatmap = (
    df.groupby(["TYPE", "YEAR"])
      .size()
      .unstack("YEAR")
      .reindex(type_order)
)

record_heatmap.columns = [
    str(int(year))
    for year in record_heatmap.columns
]

record_heatmap["Overall"] = (
    df.groupby("TYPE")
      .size()
      .reindex(type_order)
)


# -------------------------------------------------
# 4. Create Graph Objects heatmap
# -------------------------------------------------

fig2 = go.Figure(
    data=go.Heatmap(
        x=nps_heatmap.columns,
        y=nps_heatmap.index,
        z=nps_heatmap.values,

        text=nps_heatmap.values,
        texttemplate="%{text:.2f}",
        textfont=dict(
            size=16,
            color="#17324D"
        ),

        customdata=record_heatmap.values,

        zmin=7,
        zmax=9,

        colorscale=[
            [0.00, "#E76F51"],
            [0.25, "#F4A261"],
            [0.50, "#F6D365"],
            [0.75, "#82C9A5"],
            [1.00, "#2A9D8F"]
        ],

        colorbar=dict(
            title="Average NPS<br>Rating",
            tickvals=[7, 7.5, 8, 8.5, 9],
            ticktext=["7.0", "7.5", "8.0", "8.5", "9.0+"],
            thickness=18,
            len=0.85
        ),

        xgap=5,
        ygap=5,

        hovertemplate=(
            "<b>Client Type: %{y}</b><br>"
            "Period: %{x}<br>"
            "Average NPS Rating: %{z:.2f} / 10<br>"
            "Records Used: %{customdata}"
            "<extra></extra>"
        )
    )
)

# -------------------------------------------------
# 5. Format graph
# -------------------------------------------------

fig2.update_layout(
    title=dict(
        text="Average NPS Rating by Client Type and Year",
        x=0.5,
        xanchor="center"
    ),

    xaxis=dict(
        title="Period",
        side="bottom",
        type="category"
    ),

    yaxis=dict(
        title="Client Type",
        autorange="reversed",
        type="category"
    ),

    template="plotly_white",

    hoverlabel=dict(
        bgcolor="white",
        font_size=13
    ),

    margin=dict(
        l=100,
        r=120,
        t=90,
        b=80
    ),

    height=430
)


# -------------------------------------------------
# 6. Display graph
# -------------------------------------------------

fig2.show(
    config={
        "displayModeBar": True,
        "displaylogo": False,
        "responsive": True,

        "toImageButtonOptions": {
            "format": "png",
            "filename": "nps_rating_heatmap",
            "scale": 2
        }
    }
)

### Key Findings from Visualisation 2

#### Overall NPS Performance

NPO clients have the highest overall average NPS Rating at **8.12**, followed very closely by Government clients at **8.11**. The difference of only 0.01 means that their overall results are practically similar.

Private clients record a lower overall average of **7.81**. Although this is still a generally positive rating, it suggests that Private clients may be less willing to recommend the company compared with Government and NPO clients.

#### Trends by Client Type

**Government clients**

The average NPS Rating decreases slightly from **7.67 in 2021** to **7.56 in 2023**. It then improves to **8.00 in 2024** and rises sharply to **8.65 in 2025**.

The 2025 result is the highest individual value in the heatmap, suggesting a strong recent improvement in customer advocacy among Government clients.

**NPO clients**

NPO ratings fluctuate more across the period. The rating rises from **7.50 in 2021** to **8.33 in 2022**, falls to **7.83 in 2023** and **7.80 in 2024**, before reaching its highest value of **8.54 in 2025**.

Although the yearly results are inconsistent, NPO clients achieve the highest overall average NPS Rating.

**Private clients**

Private clients show the most consistent improvement. Their average NPS Rating increases every year from **7.38 in 2021** to **8.02 in 2025**.

Despite having the lowest overall average, the continuous improvement suggests that Private clients’ willingness to recommend the company is gradually strengthening.

#### Overall Trend

The year **2025 is the strongest period for all three client types**:

* Government: **8.65**
* NPO: **8.54**
* Private: **8.02**

This shared improvement may indicate that the overall client experience became more positive in 2025. The company should investigate whether any service improvements, project practices or relationship-management strategies were introduced during this period.

However, the chart alone cannot determine what caused the increase.

#### Business Implications

The company should:

1. Identify and maintain the practices that contributed to the strong 2025 results.
2. Continue improving the Private client experience, as this segment has the lowest overall NPS Rating.
3. Maintain its strong relationships with Government and NPO clients.
4. Examine why NPO ratings fluctuate more between years to determine whether their experience is inconsistent.
5. Convert satisfied clients into stronger advocates, as the overall averages remain below the Promoter threshold of 9.

#### Limitations

The chart displays **average NPS Ratings**, not the formal Net Promoter Score. An average rating cannot show the actual percentages of Promoters, Passives and Detractors within each segment.

The client groups may also contain different numbers of records, meaning that averages from smaller groups may be more affected by individual responses.

Finally, the colour scale focuses on ratings from approximately 7 to 9. This makes differences easier to see but can make small numerical gaps appear larger than they are. The printed values should therefore be considered alongside the colours.



## Visualisation 3 — Service Improvement Priority Matrix

### Purpose

This visualisation identifies how the company should prioritise its four service dimensions within each client segment:

* Presales & Partnership
* Technical Expertise
* Project Delivery
* Post-Sales Support

It evaluates each service dimension based on its current performance and its association with NPS Rating.

### How to Read the Matrix

Each point represents one service dimension.

* **X-axis — Performance:** The average service rating from 1 to 5. A point further to the right indicates higher current satisfaction.
* **Y-axis — Importance:** The Spearman correlation between the service dimension and NPS Rating. A point higher on the chart indicates a stronger association with customer advocacy.

The position of a service dimension indicates both how well it is currently performing and how closely it is related to NPS Rating.

### Priority Quadrants

The matrix is divided into four action areas:

| Quadrant             | Interpretation                                | Suggested Action                            |
| -------------------- | --------------------------------------------- | ------------------------------------------- |
| **Improve First**    | Lower performance and higher NPS association  | Prioritise improvement                      |
| **Protect Strength** | Higher performance and higher NPS association | Maintain and protect quality                |
| **Monitor**          | Lower performance and lower NPS association   | Observe and address after higher priorities |
| **Maintain**         | Higher performance and lower NPS association  | Maintain efficiently                        |

A service dimension in **Improve First** is relatively weak but more closely associated with NPS, making it an important improvement opportunity.

A dimension in **Protect Strength** is both highly rated and strongly associated with NPS. Its quality should be maintained because deterioration may negatively affect customer advocacy.

### Calculation Method

For each service dimension:

* Performance is calculated using its average service rating.
* Importance is calculated using its Spearman correlation with NPS Rating.
* Record count represents the number of valid client-year observations used.

The vertical and horizontal dividing lines represent the median performance and median correlation of the four dimensions within the selected segment.

As these dividing lines change between segments, the quadrants show **relative priorities within each segment** rather than universal service standards.

### How to Interact

* Use the **Client Segment** dropdown to switch between Overall, Government, NPO and Private clients.
* Hover over a point to view the average rating, Spearman correlation, number of records and recommended action.
* Compare the position of the four dimensions to determine which services should be improved, protected, maintained or monitored.
* Use the Plotly toolbar to zoom, pan, reset or download the visualisation.

### Interpretation Note

The matrix supports service prioritisation, but it does not prove that improving a particular dimension will directly increase NPS Rating. Spearman correlation measures association rather than causation.

The recommendations are also relative to the four service dimensions in each segment. A dimension classified as **Monitor** may still require improvement even though it is not the highest immediate priority.





In [16]:
from plotly.graph_objects import Figure, Scatter
import numpy as np
import pandas as pd


def kay_fig3(d, segment="Overall"):

    # -------------------------------------------------
    # 1. Apply client segment
    # -------------------------------------------------

    if segment == "Overall":
        g = d.copy()
    else:
        g = d[d["TYPE"].eq(segment)].copy()

    if g.empty:
        return Figure().update_layout(
            title="No data match the selected filters",
            template="plotly_white",
            height=680
        )

    # -------------------------------------------------
    # 2. Create overall satisfaction measure
    # -------------------------------------------------

    individual_service_columns = [
        "PRESALES AND PARTNERSHIP",
        "TECHNICAL EXPERTISE",
        "PROJECT DELIVERY",
        "POST-SALES SUPPORT"
    ]

    g["OVERALL SERVICE RATING"] = (
        g[individual_service_columns]
        .mean(axis=1)
    )

    service_pairs = [
        (
            "OVERALL SERVICE RATING",
            "Overall Service Satisfaction"
        ),
        (
            "PRESALES AND PARTNERSHIP",
            "Presales & Partnership"
        ),
        (
            "TECHNICAL EXPERTISE",
            "Technical Expertise"
        ),
        (
            "PROJECT DELIVERY",
            "Project Delivery"
        ),
        (
            "POST-SALES SUPPORT",
            "Post-Sales Support"
        )
    ]

    type_colours = {
        "Govt": "#636EFA",
        "NPO": "#00CC96",
        "Private": "#EF553B"
    }

    type_order = [
        client_type
        for client_type in ["Govt", "NPO", "Private"]
        if client_type in g["TYPE"].dropna().unique()
    ]

    fig = Figure()

    dimension_trace_indices = {}
    correlations = {}

    rng = np.random.default_rng(42)

    # -------------------------------------------------
    # 3. Correlation interpretation function
    # -------------------------------------------------

    def interpret_correlation(value):

        if pd.isna(value):
            return "Correlation unavailable"

        absolute_value = abs(value)

        if absolute_value < 0.20:
            strength = "Very weak"
        elif absolute_value < 0.40:
            strength = "Weak"
        elif absolute_value < 0.60:
            strength = "Moderate"
        elif absolute_value < 0.80:
            strength = "Strong"
        else:
            strength = "Very strong"

        direction = "positive" if value >= 0 else "negative"

        return f"{strength} {direction} association"

    # -------------------------------------------------
    # 4. Annotation function
    # -------------------------------------------------

    def create_annotations(correlation):

        if pd.isna(correlation):
            correlation_text = (
                "<b>Spearman correlation unavailable</b>"
            )
            background_colour = "#F4F5F7"
            border_colour = "#7A869A"

        else:
            correlation_text = (
                f"<b>Spearman ρ = {correlation:+.3f}</b>"
                f"<br>{interpret_correlation(correlation)}"
            )

            if correlation >= 0:
                background_colour = "#E5F5F1"
                border_colour = "#2A9D8F"
            else:
                background_colour = "#FDECEA"
                border_colour = "#E76F51"

        return [
            # Dropdown label
            {
                "text": "<b>Select Service Dimension</b>",
                "x": 1,
                "y": 1.23,
                "xref": "paper",
                "yref": "paper",
                "showarrow": False,
                "xanchor": "right",
                "font": {
                    "size": 13,
                    "color": "#172B4D"
                }
            },

            # Prominent correlation badge
            {
                "text": correlation_text,
                "x": 0.02,
                "y": 0.97,
                "xref": "paper",
                "yref": "paper",
                "showarrow": False,
                "xanchor": "left",
                "yanchor": "top",

                "align": "left",

                "font": {
                    "size": 16,
                    "color": "#172B4D"
                },

                "bgcolor": background_colour,
                "bordercolor": border_colour,
                "borderwidth": 2,
                "borderpad": 10
            },

            # Explanation below graph
            {
                "text": (
                    "Overall Service Satisfaction is the mean of "
                    "the four service ratings. Jitter only separates "
                    "overlapping points visually."
                ),
                "x": 0.5,
                "y": -0.20,
                "xref": "paper",
                "yref": "paper",
                "showarrow": False,
                "xanchor": "center",

                "font": {
                    "size": 11,
                    "color": "#5E6C84"
                }
            }
        ]

    # -------------------------------------------------
    # 5. Generate traces
    # -------------------------------------------------

    for dimension_number, (column, label) in enumerate(
        service_pairs
    ):

        plot_data = g[
            [column, "NPS RATING", "TYPE", "YEAR"]
        ].dropna().copy()

        correlation = plot_data[column].corr(
            plot_data["NPS RATING"],
            method="spearman"
        )

        correlations[label] = correlation

        # Add visual jitter
        plot_data["Rating Jitter"] = (
            plot_data[column]
            + rng.uniform(-0.10, 0.10, len(plot_data))
        )

        plot_data["NPS Jitter"] = (
            plot_data["NPS RATING"]
            + rng.uniform(-0.14, 0.14, len(plot_data))
        )

        trace_indices = []

        # Show Overall Service Satisfaction initially
        initial_visibility = dimension_number == 0

        # Client observations
        for client_type in type_order:

            client_data = plot_data[
                plot_data["TYPE"].eq(client_type)
            ]

            if client_data.empty:
                continue

            fig.add_trace(
                Scatter(
                    x=client_data["Rating Jitter"],
                    y=client_data["NPS Jitter"],

                    mode="markers",
                    name=client_type,
                    legendgroup=client_type,
                    showlegend=True,
                    visible=initial_visibility,

                    marker={
                        "size": 10,
                        "opacity": 0.60,
                        "color": type_colours.get(
                            client_type,
                            "#7A869A"
                        ),
                        "line": {
                            "width": 0.7,
                            "color": "white"
                        }
                    },

                    customdata=client_data[
                        [
                            column,
                            "NPS RATING",
                            "TYPE",
                            "YEAR"
                        ]
                    ].to_numpy(),

                    hovertemplate=(
                        f"<b>{label}</b><br>"
                        "Client Type: %{customdata[2]}<br>"
                        "Year: %{customdata[3]:.0f}<br>"
                        "Service Rating: "
                        "%{customdata[0]:.2f} / 5<br>"
                        "NPS Rating: "
                        "%{customdata[1]:.0f} / 10"
                        "<extra></extra>"
                    )
                )
            )

            trace_indices.append(len(fig.data) - 1)

        # Calculate average NPS at each rating level
        rating_means = (
            plot_data
            .groupby(column, as_index=False)
            .agg(
                Average_NPS=("NPS RATING", "mean"),
                Records=("NPS RATING", "size")
            )
            .sort_values(column)
        )

        # Average NPS trend
        fig.add_trace(
            Scatter(
                x=rating_means[column],
                y=rating_means["Average_NPS"],

                mode="lines+markers",
                name="Average NPS by Rating",
                legendgroup="Average NPS",
                showlegend=True,
                visible=initial_visibility,

                line={
                    "color": "#172B4D",
                    "width": 4
                },

                marker={
                    "size": 11,
                    "color": "#172B4D",
                    "symbol": "diamond"
                },

                customdata=rating_means[
                    ["Records"]
                ].to_numpy(),

                hovertemplate=(
                    "Service Rating: %{x:.2f} / 5<br>"
                    "Average NPS: %{y:.2f} / 10<br>"
                    "Records: %{customdata[0]}"
                    "<extra>Average NPS</extra>"
                )
            )
        )

        trace_indices.append(len(fig.data) - 1)

        dimension_trace_indices[label] = trace_indices

    # -------------------------------------------------
    # 6. Create dropdown
    # -------------------------------------------------

    buttons = []

    for column, label in service_pairs:

        visibility = [False] * len(fig.data)

        for trace_index in dimension_trace_indices[label]:
            visibility[trace_index] = True

        correlation = correlations[label]

        buttons.append(
            {
                "label": label,
                "method": "update",
                "args": [
                    {
                        "visible": visibility
                    },
                    {
                        "title": {
                            "text": (
                                f"{label} Against NPS"
                                f" — {segment}"
                            ),
                            "x": 0.5,
                            "xanchor": "center"
                        },

                        # Updates correlation badge
                        "annotations": create_annotations(
                            correlation
                        )
                    }
                ]
            }
        )

    # -------------------------------------------------
    # 7. Initial graph layout
    # -------------------------------------------------

    initial_label = service_pairs[0][1]
    initial_correlation = correlations[initial_label]

    fig.update_layout(
        title={
            "text": (
                f"{initial_label} Against NPS"
                f" — {segment}"
            ),
            "x": 0.5,
            "xanchor": "center",
            "font": {
                "size": 23,
                "color": "#172B4D"
            }
        },

        template="plotly_white",
        height=700,
        hovermode="closest",

        # Dropdown moved to top-right
        updatemenus=[
            {
                "buttons": buttons,
                "direction": "down",
                "showactive": True,
                "active": 0,

                "x": 1,
                "xanchor": "right",
                "y": 1.16,
                "yanchor": "top",

                "pad": {
                    "r": 0,
                    "t": 5
                }
            }
        ],

        annotations=create_annotations(
            initial_correlation
        ),

        xaxis={
            "title": "Selected Service Rating (1–5)",
            "tickmode": "array",
            "tickvals": [1, 2, 3, 4, 5],
            "range": [0.65, 5.35],
            "gridcolor": "#E5E9F0"
        },

        yaxis={
            "title": "NPS Rating (0–10)",
            "tickmode": "array",
            "tickvals": [0, 2, 4, 6, 8, 10],
            "range": [-0.5, 10.5],
            "gridcolor": "#E5E9F0"
        },

        # Legend moved below the title and away from dropdown
        legend={
            "title": "Client Type",
            "orientation": "h",
            "yanchor": "bottom",
            "y": 1.02,
            "xanchor": "center",
            "x": 0.5
        },

        margin={
            "t": 180,
            "l": 80,
            "r": 40,
            "b": 125
        }
    )

    return fig


# -------------------------------------------------
# 8. Display graph
# -------------------------------------------------

fig3 = kay_fig3(
    df,
    "Overall"
)

fig3.show(
    config={
        "displayModeBar": True,
        "displaylogo": False,
        "scrollZoom": True,
        "responsive": True,

        "toImageButtonOptions": {
            "format": "png",
            "filename": "service_ratings_against_nps",
            "scale": 2
        }
    }
)

### Key Findings from Visualisation 3

#### Overall Priorities

The Overall matrix separates the four service dimensions into different action areas based on their average performance and association with NPS.

| Service Dimension          | Average Rating | NPS Correlation | Priority         |
| -------------------------- | -------------: | --------------: | ---------------- |
| **Technical Expertise**    |           3.92 |           0.544 | Improve First    |
| **Project Delivery**       |           4.08 |           0.538 | Protect Strength |
| **Post-Sales Support**     |           3.84 |           0.444 | Monitor          |
| **Presales & Partnership** |           3.96 |           0.412 | Maintain         |

**Technical Expertise** falls under Improve First. Although its average rating is generally positive, it is below the relative performance midpoint and has the strongest overall association with NPS at 0.544. This makes it a more important improvement opportunity than simply selecting the lowest-rated service dimension.

**Project Delivery** falls under Protect Strength. It has the highest average performance at 4.08 and a strong NPS association of 0.538. This suggests that Project Delivery is both well-performing and relatively important to customer advocacy.

**Post-Sales Support** falls under Monitor. It has the lowest overall average rating at 3.84, but its NPS association of 0.444 is below the relative importance midpoint. It should still be observed and improved where necessary, but the matrix does not classify it as the first overall priority.

**Presales & Partnership** falls under Maintain. Its performance is above the relative midpoint, but its NPS association of 0.412 is lower than those of Technical Expertise and Project Delivery.

#### Priorities by Client Segment

The matrix shows that different client segments require different improvement priorities.

| Client Segment | Improve First          | Protect Strength | Monitor             | Maintain               |
| -------------- | ---------------------- | ---------------- | ------------------- | ---------------------- |
| **Government** | Technical Expertise    | Project Delivery | Post-Sales Support  | Presales & Partnership |
| **NPO**        | Presales & Partnership | Project Delivery | Technical Expertise | Post-Sales Support     |
| **Private**    | Technical Expertise    | Project Delivery | Post-Sales Support  | Presales & Partnership |

#### Government Clients

For Government clients, **Technical Expertise** is classified as Improve First. Its average rating of 3.94 is below the segment’s relative performance midpoint, while its correlation with NPS is comparatively high at 0.522.

**Project Delivery** is classified as Protect Strength because it combines the highest Government rating of 4.35 with the strongest Government NPS association of 0.559.

This suggests that the company should maintain the quality of Government project execution while examining how its technical service experience can be strengthened.

#### NPO Clients

For NPO clients, **Presales & Partnership** is classified as Improve First. It has the lowest NPO service rating at 3.76 but a strong NPS association of 0.619.

This combination indicates that early communication, expectation setting and relationship-building may be particularly important improvement areas for NPO clients.

**Project Delivery** is classified as Protect Strength. It has a high average rating of 4.26 and the strongest NPO correlation with NPS at 0.670.

Technical Expertise is placed under Monitor, while Post-Sales Support is classified as Maintain. This means that NPO Presales should receive greater attention than simply improving every lower-rated service equally.

#### Private Clients

For Private clients, **Technical Expertise** is classified as Improve First. Its average rating of 3.92 is slightly below the segment’s relative performance midpoint, while it has the strongest Private NPS association at 0.571.

**Project Delivery** is classified as Protect Strength because it combines the highest Private service rating of 4.01 with a strong correlation of 0.521.

Post-Sales Support is classified as Monitor. Although it is the weakest-rated Private dimension at 3.80, its association with NPS is lower than those of Technical Expertise and Project Delivery. It remains a service weakness, but the matrix suggests that Technical Expertise may be the more important immediate priority for strengthening advocacy.

#### Main Strategic Finding

The most consistent result is that **Project Delivery should be protected across all three client segments**. It appears in the Protect Strength quadrant for Government, NPO and Private clients.

The main improvement priority differs by segment:

* Improve Technical Expertise for Government clients.
* Improve Presales & Partnership for NPO clients.
* Improve Technical Expertise for Private clients.

This supports a segment-specific approach instead of applying the same improvement programme to every client type.

#### Business Implications

The company should:

1. Protect Project Delivery quality across all client segments.
2. Strengthen Technical Expertise for Government and Private clients.
3. Improve early engagement and relationship-building for NPO clients.
4. Continue monitoring Post-Sales Support, especially where individual feedback indicates service problems.
5. Use both current performance and NPS association when allocating improvement resources.

#### Limitations

The quadrant boundaries are based on the median values of only four service dimensions. A small change in a rating or correlation could therefore move a dimension into a different quadrant.

The classifications are also relative to each selected segment. For example, Monitor does not mean that a service is satisfactory or unimportant; it only indicates that it is a lower priority compared with the other dimensions in that segment.

Average ratings may hide differences between individual clients, and correlations may be less stable for segments with fewer observations, particularly NPO clients.

Finally, correlation measures association rather than causation. The matrix identifies evidence-based priorities but does not prove that improving a particular service dimension will directly increase NPS.



## Interactive Satisfaction Dashboard

The dashboard combines the three visualisations into one decision-making view:

1. The **grouped bar chart** compares the four service dimensions across client types and identifies current service strengths and weaknesses.
2. The **NPS heatmap** compares customer advocacy across client types and years, highlighting changes over time.
3. The **Service Improvement Priority Matrix** combines average service performance with its association with NPS to determine which dimensions should be improved, protected, maintained or monitored.

The KPI cards provide a summary of the currently filtered data, including:

* Average Service Rating
* Average NPS Rating
* Strongest Service Dimension
* Number of Filtered Records

The navigation filters allow users to explore the dashboard by client type, year and service dimension. When a filter is changed, the KPI cards and visualisations update to reflect the selected records.

Together, the dashboard guides users through three stages:

> **How are services performing? → How does customer advocacy vary? → Which service areas should be prioritised?**

Run the cell below to launch the dashboard locally.


In [17]:
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

from dash import Dash, Input, Output, dcc, html


# -------------------------------------------------
# 1. Dashboard data and labels
# -------------------------------------------------

if "df" in globals():
    dashboard_df = df.copy()
else:
    dashboard_df = pd.read_excel("merged_cleaned.xlsx")
dashboard_df["YEAR"] = pd.to_numeric(
    dashboard_df["YEAR"], errors="coerce"
).astype("Int64")

service_columns = [
    "PRESALES AND PARTNERSHIP",
    "TECHNICAL EXPERTISE",
    "PROJECT DELIVERY",
    "POST-SALES SUPPORT"
]

service_label_map = {
    "PRESALES AND PARTNERSHIP": "Presales & Partnership",
    "TECHNICAL EXPERTISE": "Technical Expertise",
    "PROJECT DELIVERY": "Project Delivery",
    "POST-SALES SUPPORT": "Post-Sales Support"
}

service_colour_map = {
    "Presales & Partnership": "#636EFA",
    "Technical Expertise": "#EF553B",
    "Project Delivery": "#00CC96",
    "Post-Sales Support": "#AB63FA"
}

client_types = sorted(dashboard_df["TYPE"].dropna().unique())
years = sorted(dashboard_df["YEAR"].dropna().astype(int).unique())


# -------------------------------------------------
# 2. Reusable styles
# -------------------------------------------------

page_style = {
    "backgroundColor": "#F4F6FA",
    "minHeight": "100vh",
    "padding": "24px",
    "fontFamily": "Segoe UI, Arial, sans-serif",
    "color": "#172B4D"
}

content_style = {
    "maxWidth": "1500px",
    "margin": "0 auto"
}

navbar_style = {
    "position": "sticky",
    "top": "0",
    "zIndex": "1000",
    "display": "grid",
    "gridTemplateColumns": "repeat(auto-fit, minmax(190px, 1fr))",
    "gap": "14px",
    "alignItems": "end",
    "backgroundColor": "#172B4D",
    "padding": "18px",
    "borderRadius": "12px",
    "boxShadow": "0 3px 12px rgba(0, 0, 0, 0.15)",
    "marginBottom": "18px"
}

filter_label_style = {
    "display": "block",
    "color": "white",
    "fontSize": "13px",
    "fontWeight": "600",
    "marginBottom": "6px"
}

metric_grid_style = {
    "display": "grid",
    "gridTemplateColumns": "repeat(auto-fit, minmax(210px, 1fr))",
    "gap": "16px",
    "marginBottom": "20px"
}

metric_card_style = {
    "backgroundColor": "white",
    "padding": "18px",
    "borderRadius": "12px",
    "boxShadow": "0 2px 8px rgba(0, 0, 0, 0.08)"
}

chart_card_style = {
    "backgroundColor": "white",
    "padding": "16px",
    "borderRadius": "12px",
    "boxShadow": "0 2px 8px rgba(0, 0, 0, 0.08)",
    "minWidth": "0",
    "overflow": "visible"
}

bottom_grid_style = {
    "display": "grid",
    "gridTemplateColumns": "1fr",
    "gap": "24px",
    "marginTop": "20px",
    "width": "100%"
}

graph_config = {
    "displaylogo": False,
    "responsive": True,
    "scrollZoom": True,
    "toImageButtonOptions": {
        "format": "png",
        "filename": "client_satisfaction_dashboard",
        "scale": 2
    }
}


# -------------------------------------------------
# 3. Figure-building functions
# -------------------------------------------------

def make_empty_figure(message):
    empty_figure = go.Figure()
    empty_figure.add_annotation(
        text=message,
        x=0.5,
        y=0.5,
        xref="paper",
        yref="paper",
        showarrow=False,
        font={"size": 16, "color": "#5E6C84"}
    )
    empty_figure.update_layout(
        template="plotly_white",
        xaxis={"visible": False},
        yaxis={"visible": False},
        height=420
    )
    return empty_figure


def make_service_figure(filtered_df, selected_services, filter_text):
    service_summary = (
        filtered_df
        .groupby("TYPE", observed=True)[selected_services]
        .mean()
        .reset_index()
    )

    record_counts = filtered_df.groupby("TYPE", observed=True).size()

    service_long = service_summary.melt(
        id_vars="TYPE",
        value_vars=selected_services,
        var_name="Service Column",
        value_name="Average Rating"
    )

    service_long["Service Dimension"] = service_long[
        "Service Column"
    ].map(service_label_map)

    service_long["Records"] = service_long["TYPE"].map(record_counts)

    figure = px.bar(
        service_long,
        x="TYPE",
        y="Average Rating",
        color="Service Dimension",
        barmode="group",
        text="Average Rating",
        custom_data=["Records"],
        color_discrete_map=service_colour_map,
        category_orders={
            "TYPE": client_types,
            "Service Dimension": [
                service_label_map[column]
                for column in selected_services
            ]
        },
        title=f"Service Ratings — {filter_text}"
    )

    figure.update_traces(
        texttemplate="%{y:.2f}",
        textposition="outside",
        cliponaxis=False,
        hovertemplate=(
            "Client Type: %{x}<br>"
            "Average Rating: %{y:.2f}<br>"
            "Records: %{customdata[0]}"
            "<extra>%{fullData.name}</extra>"
        )
    )

    figure.update_layout(
        template="plotly_white",
        height=560,
        yaxis={"range": [0, 5], "title": "Average Rating"},
        xaxis_title="Client Type",
        legend_title="Service Dimension",
        title_x=0.5,
        margin={"t": 80, "r": 30, "b": 60, "l": 60}
    )

    return figure


def make_nps_figure(filtered_df, selected_year, filter_text):
    nps_summary = (
        filtered_df
        .groupby(["TYPE", "YEAR"], observed=True)["NPS RATING"]
        .agg(["mean", "count"])
        .reset_index()
    )

    value_matrix = nps_summary.pivot(
        index="TYPE",
        columns="YEAR",
        values="mean"
    )

    count_matrix = nps_summary.pivot(
        index="TYPE",
        columns="YEAR",
        values="count"
    )

    displayed_types = [
        client_type
        for client_type in client_types
        if client_type in value_matrix.index
    ]

    if selected_year == "All":
        displayed_periods = [
            year for year in years if year in value_matrix.columns
        ]

        overall_nps = filtered_df.groupby(
            "TYPE", observed=True
        )["NPS RATING"].mean()

        overall_count = filtered_df.groupby(
            "TYPE", observed=True
        ).size()

        value_matrix["Overall"] = overall_nps
        count_matrix["Overall"] = overall_count
        displayed_periods = displayed_periods + ["Overall"]
    else:
        displayed_periods = [int(selected_year)]

    value_matrix = value_matrix.reindex(
        index=displayed_types,
        columns=displayed_periods
    )

    count_matrix = count_matrix.reindex(
        index=displayed_types,
        columns=displayed_periods
    )

    heatmap_text = np.array([
        ["" if pd.isna(value) else f"{value:.2f}" for value in row]
        for row in value_matrix.to_numpy()
    ])

    figure = go.Figure(
        data=go.Heatmap(
            z=value_matrix.to_numpy(),
            x=[str(period) for period in displayed_periods],
            y=displayed_types,
            text=heatmap_text,
            texttemplate="%{text}",
            textfont={"size": 16},
            customdata=count_matrix.to_numpy(),
            colorscale=[
                [0.00, "#E76F51"],
                [0.50, "#F4D35E"],
                [1.00, "#2A9D8F"]
            ],
            zmin=7,
            zmax=9,
            colorbar={"title": "Average<br>NPS Rating"},
            hovertemplate=(
                "Client Type: %{y}<br>"
                "Period: %{x}<br>"
                "Average NPS Rating: %{z:.2f}<br>"
                "Records: %{customdata:.0f}"
                "<extra></extra>"
            )
        )
    )

    figure.update_layout(
        template="plotly_white",
        title=f"Average NPS Rating — {filter_text}",
        title_x=0.5,
        xaxis_title="Period",
        yaxis_title="Client Type",
        height=470,
        margin={"t": 80, "r": 30, "b": 60, "l": 70}
    )

    return figure


def make_priority_figure(filtered_df, selected_services, filter_text):
    """Combine current service performance and NPS association."""

    priority_rows = []

    for column in selected_services:
        pair = filtered_df[[column, "NPS RATING"]].dropna()

        enough_variation = (
            len(pair) >= 3
            and pair[column].nunique() >= 2
            and pair["NPS RATING"].nunique() >= 2
        )

        correlation = (
            pair[column].corr(pair["NPS RATING"], method="spearman")
            if enough_variation
            else np.nan
        )

        priority_rows.append({
            "Service Dimension": service_label_map[column],
            "Performance": pair[column].mean(),
            "Importance": correlation,
            "Records": len(pair)
        })

    priority_df = pd.DataFrame(priority_rows).dropna(
        subset=["Performance", "Importance"]
    )

    if len(priority_df) < 2:
        return make_empty_figure(
            "Select at least two service dimensions with sufficient data."
        )

    performance_cutoff = priority_df["Performance"].median()
    importance_cutoff = priority_df["Importance"].median()

    def assign_priority(row):
        high_performance = row["Performance"] >= performance_cutoff
        high_importance = row["Importance"] >= importance_cutoff

        if not high_performance and high_importance:
            return "Improve First"
        if high_performance and high_importance:
            return "Protect Strength"
        if not high_performance and not high_importance:
            return "Monitor"
        return "Maintain"

    priority_df["Recommended Action"] = priority_df.apply(
        assign_priority,
        axis=1
    )

    x_span = priority_df["Performance"].max() - priority_df["Performance"].min()
    y_span = priority_df["Importance"].max() - priority_df["Importance"].min()

    x_padding = max(0.18, x_span * 0.40)
    y_padding = max(0.10, y_span * 0.40)

    x_min = max(1.0, priority_df["Performance"].min() - x_padding)
    x_max = min(5.0, priority_df["Performance"].max() + x_padding)
    y_min = max(-1.0, priority_df["Importance"].min() - y_padding)
    y_max = min(1.0, priority_df["Importance"].max() + y_padding)

    # Keep the dividing lines away from the chart edges when values are equal.
    if x_min == x_max:
        x_min, x_max = max(1.0, x_min - 0.25), min(5.0, x_max + 0.25)
    if y_min == y_max:
        y_min, y_max = max(-1.0, y_min - 0.15), min(1.0, y_max + 0.15)

    action_colours = {
        "Improve First": "#E76F51",
        "Protect Strength": "#2A9D8F",
        "Monitor": "#F4A261",
        "Maintain": "#457B9D"
    }

    text_positions = {
        "Presales & Partnership": "bottom left",
        "Technical Expertise": "top left",
        "Project Delivery": "top right",
        "Post-Sales Support": "bottom right"
    }

    figure = go.Figure()

    quadrant_specs = [
        (
            x_min, performance_cutoff,
            importance_cutoff, y_max,
            "#FDE7E2", "Improve First",
            x_min, y_max, "left", "top"
        ),
        (
            performance_cutoff, x_max,
            importance_cutoff, y_max,
            "#DDF2EC", "Protect Strength",
            x_max, y_max, "right", "top"
        ),
        (
            x_min, performance_cutoff,
            y_min, importance_cutoff,
            "#FFF1DB", "Monitor",
            x_min, y_min, "left", "bottom"
        ),
        (
            performance_cutoff, x_max,
            y_min, importance_cutoff,
            "#E6EEF7", "Maintain",
            x_max, y_min, "right", "bottom"
        )
    ]

    for (
        quadrant_x0, quadrant_x1, quadrant_y0, quadrant_y1,
        fill_colour, label, label_x, label_y, x_anchor, y_anchor
    ) in quadrant_specs:
        figure.add_shape(
            type="rect",
            x0=quadrant_x0,
            x1=quadrant_x1,
            y0=quadrant_y0,
            y1=quadrant_y1,
            fillcolor=fill_colour,
            opacity=0.55,
            line_width=0,
            layer="below"
        )
        figure.add_annotation(
            x=label_x,
            y=label_y,
            text=f"<b>{label}</b>",
            showarrow=False,
            xanchor=x_anchor,
            yanchor=y_anchor,
            xshift=12 if x_anchor == "left" else -12,
            yshift=-10 if y_anchor == "top" else 10,
            font={"size": 13, "color": "#42526E"}
        )

    for action in [
        "Improve First", "Protect Strength", "Monitor", "Maintain"
    ]:
        action_data = priority_df[
            priority_df["Recommended Action"].eq(action)
        ]

        if action_data.empty:
            continue

        custom_data = action_data[
            ["Records", "Recommended Action"]
        ].to_numpy(dtype=object)

        figure.add_trace(
            go.Scatter(
                x=action_data["Performance"],
                y=action_data["Importance"],
                mode="markers+text",
                name=action,
                text=[
                    (
                        f"<b>{service}</b><br>"
                        f"{performance:.2f} | ρ {importance:+.2f}"
                    )
                    for service, performance, importance in zip(
                        action_data["Service Dimension"],
                        action_data["Performance"],
                        action_data["Importance"]
                    )
                ],
                textposition=[
                    text_positions.get(service, "top center")
                    for service in action_data["Service Dimension"]
                ],
                marker={
                    "size": 19,
                    "color": action_colours[action],
                    "line": {"width": 2, "color": "white"}
                },
                customdata=custom_data,
                hovertemplate=(
                    "<b>%{text}</b><br>"
                    "Average Rating: %{x:.3f} / 5<br>"
                    "Spearman Correlation: %{y:+.3f}<br>"
                    "Valid Records: %{customdata[0]}<br>"
                    "Recommended Action: %{customdata[1]}"
                    "<extra></extra>"
                )
            )
        )

    figure.add_vline(
        x=performance_cutoff,
        line_dash="dash",
        line_color="#5E6C84",
        line_width=2
    )
    figure.add_hline(
        y=importance_cutoff,
        line_dash="dash",
        line_color="#5E6C84",
        line_width=2
    )

    figure.add_annotation(
        x=performance_cutoff,
        y=y_min,
        text=f"Performance median: {performance_cutoff:.2f}",
        showarrow=False,
        xanchor="center",
        yanchor="bottom",
        yshift=6,
        font={"size": 11, "color": "#5E6C84"},
        bgcolor="rgba(255,255,255,0.80)"
    )
    figure.add_annotation(
        x=x_min,
        y=importance_cutoff,
        text=f"Importance median: {importance_cutoff:+.2f}",
        showarrow=False,
        xanchor="left",
        yanchor="bottom",
        xshift=8,
        yshift=4,
        font={"size": 11, "color": "#5E6C84"},
        bgcolor="rgba(255,255,255,0.80)"
    )

    figure.update_layout(
        template="plotly_white",
        title={
            "text": (
                f"Service Improvement Priority Matrix — {filter_text}"
                "<br><sup>"
                "Performance = average rating; Importance = Spearman "
                "association with NPS Rating"
                "</sup>"
            ),
            "x": 0.5,
            "xanchor": "center"
        },
        xaxis={
            "title": "Current Performance — Average Service Rating (1–5)",
            "range": [x_min, x_max],
            "tickformat": ".2f",
            "showgrid": True
        },
        yaxis={
            "title": "Relative Importance — Spearman Correlation with NPS",
            "range": [y_min, y_max],
            "tickformat": ".2f",
            "showgrid": True
        },
        height=680,
        hovermode="closest",
        legend={
            "title": "Recommended Action",
            "orientation": "h",
            "yanchor": "bottom",
            "y": 1.02,
            "xanchor": "center",
            "x": 0.5
        },
        margin={"t": 130, "r": 80, "b": 85, "l": 95}
    )

    return figure


# -------------------------------------------------
# 4. Build the Dash application
# -------------------------------------------------

app = Dash(__name__)
app.title = "Client Satisfaction Dashboard"

app.layout = html.Div(
    style=page_style,
    children=[
        html.Div(
            style=content_style,
            children=[
                html.Nav(
                    style=navbar_style,
                    children=[
                        html.Div(
                            children=[
                                html.H2(
                                    "Client Satisfaction Dashboard",
                                    style={
                                        "color": "white",
                                        "margin": "0 0 4px 0"
                                    }
                                ),
                                html.P(
                                    id="filter-summary",
                                    style={
                                        "color": "#D6E0F0",
                                        "margin": "0",
                                        "fontSize": "13px"
                                    }
                                )
                            ]
                        ),

                        html.Div(
                            children=[
                                html.Label(
                                    "Client Type",
                                    style=filter_label_style
                                ),
                                dcc.Dropdown(
                                    id="client-type-filter",
                                    options=(
                                        [{"label": "All Client Types", "value": "All"}]
                                        + [
                                            {"label": value, "value": value}
                                            for value in client_types
                                        ]
                                    ),
                                    value="All",
                                    clearable=False
                                )
                            ]
                        ),

                        html.Div(
                            children=[
                                html.Label(
                                    "Year",
                                    style=filter_label_style
                                ),
                                dcc.Dropdown(
                                    id="year-filter",
                                    options=(
                                        [{"label": "All Years", "value": "All"}]
                                        + [
                                            {"label": str(year), "value": year}
                                            for year in years
                                        ]
                                    ),
                                    value="All",
                                    clearable=False
                                )
                            ]
                        ),

                        html.Div(
                            children=[
                                html.Label(
                                    "Service Dimensions (select 2+ for matrix)",
                                    style=filter_label_style
                                ),
                                dcc.Dropdown(
                                    id="service-filter",
                                    options=[
                                        {
                                            "label": service_label_map[column],
                                            "value": column
                                        }
                                        for column in service_columns
                                    ],
                                    value=service_columns,
                                    multi=True,
                                    clearable=False
                                )
                            ]
                        ),

                        html.Button(
                            "Reset Filters",
                            id="reset-filters",
                            n_clicks=0,
                            style={
                                "height": "38px",
                                "padding": "0 18px",
                                "border": "1px solid #D6E0F0",
                                "borderRadius": "6px",
                                "backgroundColor": "white",
                                "color": "#172B4D",
                                "fontWeight": "600",
                                "cursor": "pointer"
                            }
                        )
                    ]
                ),

                html.Div(
                    style=metric_grid_style,
                    children=[
                        html.Div(
                            style=metric_card_style,
                            children=[
                                html.P(
                                    "Average Service Rating",
                                    style={"color": "#5E6C84", "margin": "0"}
                                ),
                                html.H2(
                                    id="overall-service-value",
                                    style={"marginBottom": "0"}
                                )
                            ]
                        ),

                        html.Div(
                            style=metric_card_style,
                            children=[
                                html.P(
                                    "Average NPS Rating",
                                    style={"color": "#5E6C84", "margin": "0"}
                                ),
                                html.H2(
                                    id="average-nps-value",
                                    style={"marginBottom": "0"}
                                )
                            ]
                        ),

                        html.Div(
                            style=metric_card_style,
                            children=[
                                html.P(
                                    "Strongest Selected Dimension",
                                    style={"color": "#5E6C84", "margin": "0"}
                                ),
                                html.H3(
                                    id="strongest-service-value",
                                    style={"marginBottom": "0"}
                                )
                            ]
                        ),

                        html.Div(
                            style=metric_card_style,
                            children=[
                                html.P(
                                    "Filtered Records",
                                    style={"color": "#5E6C84", "margin": "0"}
                                ),
                                html.H2(
                                    id="record-count-value",
                                    style={"marginBottom": "0"}
                                )
                            ]
                        )
                    ]
                ),

                html.Div(
                    style=chart_card_style,
                    children=[
                        dcc.Loading(
                            dcc.Graph(
                                id="service-rating-chart",
                                config=graph_config,
                                responsive=True
                            )
                        )
                    ]
                ),

                html.Div(
                    style=bottom_grid_style,
                    children=[
                        html.Div(
                            style=chart_card_style,
                            children=[
                                dcc.Loading(
                                    dcc.Graph(
                                        id="nps-heatmap",
                                        config=graph_config,
                                        responsive=True
                                    )
                                )
                            ]
                        ),

                        html.Div(
                            style=chart_card_style,
                            children=[
                                dcc.Loading(
                                    dcc.Graph(
                                        id="service-priority-matrix",
                                        config=graph_config,
                                        responsive=True,
                                        style={
                                            "height": "680px",
                                            "width": "100%"
                                        }
                                    )
                                )
                            ]
                        )
                    ]
                ),

                html.P(
                    (
                        "Priority quadrants are relative to the selected data. "
                        "Correlations indicate association, not causation."
                    ),
                    style={
                        "textAlign": "center",
                        "color": "#7A869A",
                        "fontSize": "12px",
                        "marginTop": "18px"
                    }
                )
            ]
        )
    ]
)


# -------------------------------------------------
# 5. Reset the navbar filters
# -------------------------------------------------

@app.callback(
    Output("client-type-filter", "value"),
    Output("year-filter", "value"),
    Output("service-filter", "value"),
    Input("reset-filters", "n_clicks"),
    prevent_initial_call=True
)
def reset_filters(_):
    return "All", "All", service_columns


# -------------------------------------------------
# 6. Update the complete dashboard
# -------------------------------------------------

@app.callback(
    Output("service-rating-chart", "figure"),
    Output("nps-heatmap", "figure"),
    Output("service-priority-matrix", "figure"),
    Output("overall-service-value", "children"),
    Output("average-nps-value", "children"),
    Output("strongest-service-value", "children"),
    Output("record-count-value", "children"),
    Output("filter-summary", "children"),
    Input("client-type-filter", "value"),
    Input("year-filter", "value"),
    Input("service-filter", "value")
)
def update_dashboard(selected_client_type, selected_year, selected_services):
    selected_services = [
        column
        for column in (selected_services or service_columns)
        if column in service_columns
    ]

    filtered_df = dashboard_df.copy()

    if selected_client_type != "All":
        filtered_df = filtered_df[
            filtered_df["TYPE"] == selected_client_type
        ]

    if selected_year != "All":
        filtered_df = filtered_df[
            filtered_df["YEAR"] == int(selected_year)
        ]

    client_text = (
        "All Client Types"
        if selected_client_type == "All"
        else selected_client_type
    )

    year_text = (
        "All Years"
        if selected_year == "All"
        else str(selected_year)
    )

    filter_text = f"{client_text}, {year_text}"

    if filtered_df.empty:
        empty = make_empty_figure("No records match the selected filters.")
        return (
            empty,
            empty,
            empty,
            "No data",
            "No data",
            "No data",
            "0",
            f"Showing: {filter_text}"
        )

    service_figure = make_service_figure(
        filtered_df,
        selected_services,
        filter_text
    )

    nps_figure = make_nps_figure(
        filtered_df,
        selected_year,
        filter_text
    )

    priority_figure = make_priority_figure(
        filtered_df,
        selected_services,
        filter_text
    )

    average_service = (
        filtered_df[selected_services]
        .mean(axis=1)
        .mean()
    )

    average_nps = filtered_df["NPS RATING"].mean()
    service_means = filtered_df[selected_services].mean()
    strongest_column = service_means.idxmax()
    strongest_name = service_label_map[strongest_column]
    strongest_rating = service_means.loc[strongest_column]

    return (
        service_figure,
        nps_figure,
        priority_figure,
        f"{average_service:.2f} / 5",
        f"{average_nps:.2f} / 10",
        f"{strongest_name} ({strongest_rating:.2f} / 5)",
        f"{len(filtered_df):,}",
        f"Showing: {filter_text}"
    )


# -------------------------------------------------
# 7. Run the dashboard
# -------------------------------------------------

if __name__ == "__main__":
    app.run(debug=True)

In [18]:
app.run(
    jupyter_mode="external",
    debug=False,
    port=8050
)

Dash app running on http://127.0.0.1:8050/


## Conclusion and Strategic Recommendations

### Overall Analytical Conclusion

The three visualisations suggest that customer satisfaction is generally positive, but the factors associated with loyalty differ across client segments. Therefore, the company should not simply focus on the lowest rating or apply the same improvement strategy to every client.

A more useful approach is to consider two factors together:

1. **Performance:** Whether the service dimension currently has a high or low average rating.
2. **Importance:** How strongly that dimension is associated with NPS Rating.

A service dimension with a **low rating and strong NPS association** represents a more urgent improvement opportunity. A dimension with a **high rating and strong association** represents a strength that should be protected.

### Government Clients: Protect a Strong Delivery Experience

Government clients report the highest overall ratings for Presales & Partnership, Technical Expertise and Project Delivery. Their average NPS Rating is also strong at **8.11**, reaching **8.65 in 2025**.

Project Delivery is both:

* the highest-rated Government service dimension at **4.35**; and
* the dimension most strongly associated with Government NPS Rating at **0.559**.

This combination suggests that Project Delivery is not simply performing well-it is also closely connected to how Government clients evaluate their overall relationship with the company.

Therefore, Project Delivery should be treated as a strength to protect rather than an area requiring major corrective action. The company should investigate what contributed to the Government Project Delivery rating increasing from **3.67 in 2021 to 4.75 in 2025**, and determine whether these practices can be standardised.

Government Post-Sales Support is lower at **3.91**, but its NPS association is also weaker than Project Delivery. It remains an improvement opportunity, although it may not have the same strategic importance as maintaining delivery quality.

### NPO Clients: Strong Outcomes but a Weak Initial Experience

NPO clients have the highest overall average NPS Rating at **8.12**, although this is only 0.01 above Government clients. They also report strong Project Delivery and Post-Sales Support ratings of **4.26** and **4.12**.

However, Presales & Partnership is their weakest service dimension at **3.76** and has a strong correlation of **0.619** with NPS Rating.

This creates an important finding: NPO clients are currently positive about the company overall, but the initial relationship-building stage may be a hidden weakness. Because Presales & Partnership is strongly associated with their NPS Rating, continued weakness in this area could prevent the company from building stronger advocacy among NPO clients.

NPO Presales & Partnership should therefore be treated as a high-priority improvement area. Actions could include clearer expectation setting, more regular early-stage communication and greater involvement of NPO clients during solution planning.

NPO ratings also fluctuate more noticeably between years. This may indicate an inconsistent experience, although the result could also be influenced by a smaller number of records. The company should examine individual NPO responses before making large operational changes.

### Private Clients: Loyalty Gap May Extend Beyond Service Ratings

Private clients have the lowest overall average NPS Rating at **7.81**, but their service ratings are not dramatically lower than those of the other segments. Their Project Delivery rating, for example, is still positive at **4.01**.

This creates a possible gap: Private clients appear reasonably satisfied with the measured service dimensions, but they are less willing to recommend the company.

Technical Expertise has the strongest association with Private NPS Rating at **0.571**, while Post-Sales Support has the lowest average rating at **3.80** and a positive association of **0.464**.

The company should therefore use a two-part strategy for Private clients:

1. Maintain and strengthen Technical Expertise because it is most closely associated with their advocacy.
2. Improve Post-Sales Support because it is their weakest reported service experience.

However, the lower Private NPS Rating may not be fully explained by these four service dimensions. Other factors not included in the dataset—such as price, perceived value, product suitability or competitor alternatives—may also influence their willingness to recommend the company. Further investigation is needed before assuming that service improvements alone will remove the loyalty gap.

### Company-Wide Pattern in 2025

All three client types record their highest NPS Rating in 2025. Project Delivery also reaches its highest value for every client type during the same year.

Because the improvement occurs across multiple segments, it may reflect a company-wide change rather than a segment-specific event. Possible explanations could include improved project processes, staff capability or changes in the clients surveyed.

The chart cannot confirm which explanation is correct. The company should compare its 2025 operational practices and client composition with previous years before treating the increase as a lasting improvement.

### Strategic Priorities

Based on both average performance and NPS association, the recommended priorities are:

| Priority                                 | Evidence                                                                      | Recommended Response                                                           |
| ---------------------------------------- | ----------------------------------------------------------------------------- | ------------------------------------------------------------------------------ |
| **Protect Project Delivery**             | Highest-rated dimension across every segment and strongly associated with NPS | Maintain quality and investigate successful 2025 practices                     |
| **Improve NPO Presales & Partnership**   | Low rating of 3.76 and strong NPS association of 0.619                        | Strengthen early communication and relationship-building                       |
| **Improve Private Post-Sales Support**   | Lowest Private service rating at 3.80                                         | Improve follow-up, support access and issue resolution                         |
| **Maintain Private Technical Expertise** | Strongest Private NPS association at 0.571                                    | Protect technical quality and communicate technical value clearly              |
| **Investigate the 2025 improvement**     | Highest NPS and Project Delivery ratings across all segments                  | Identify whether operational changes or sample composition caused the increase |

### Connection to the Main Problem

This analysis identifies which client segments maintain stronger customer satisfaction, but it cannot independently determine which segments should receive the highest business priority.

The satisfaction findings should be combined with the team’s profitability results using the following framework:

* **High profit and high satisfaction:** Protect and grow the relationship.
* **High profit and low satisfaction:** Prioritise service recovery because valuable clients may be at risk.
* **Low profit and high satisfaction:** Explore cross-selling, pricing or cost optimisation opportunities.
* **Low profit and low satisfaction:** Reassess the commercial value of continued investment.

Government and NPO clients currently show stronger overall satisfaction and advocacy, while Private clients present clearer improvement opportunities. However, the final recommendation depends on whether these segments also generate sufficient gross profit and gross margin.

### Final Limitation

These findings identify patterns and associations rather than proven causes. Average ratings may hide differences between individual clients, the client segments have unequal sample sizes, and some clients appear across multiple years.

The analysis also uses average NPS Rating rather than the formal Net Promoter Score and does not contain actual renewal or churn outcomes. Therefore, NPS should be interpreted as an indicator of customer advocacy—not proof of future retention.

The results provide evidence for prioritisation, but the company should validate the findings using client-level feedback, sample sizes, profitability measures and actual retention outcomes before making major strategic decisions.
